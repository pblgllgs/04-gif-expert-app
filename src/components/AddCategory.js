import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
    const [inputValue, setInputValue] = useState("");

    const handleInputchange = (e) => {
        setInputValue(e.target.value);
        console.log('HandleInputChange Llamado')
    };

    const handleSubmit = (e) => {
        e.preventDefault();
        if (inputValue.trim().length > 2) {
            setCategories(categories => [inputValue, ...categories]);
            setInputValue('');
        }
    };

    return (
        <form onSubmit={handleSubmit}>
            <p>{inputValue}</p>
            <input
                type="text"
                placeholder="Ingresa una categoria"
                value={inputValue}
                onChange={handleInputchange}
            />
        </form>
    );
};

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
};

export default AddCategory;
