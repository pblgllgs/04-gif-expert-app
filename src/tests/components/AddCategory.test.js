import React from "react";
import { shallow } from "enzyme";
import AddCategory from "../../components/AddCategory";

describe('Pruebas en <AddCategory />', () => {

    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories} />);

    beforeEach( ()=>{
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories} />);
    })

    test("Debe mostrar <AddCategory /> correctamente", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de cambiar el input', () => {
        const input =wrapper.find('input');
        const value = 'Hola';
        // simulamos que el target trae dentro un elemento con el value, sino da error
        input.simulate('change',{target : {value}});
        expect(wrapper.find('p').text().trim()).toBe(value);
    });

    test('No debe de postear la informacion con submit', () => {
        wrapper.find('form').simulate('submit',{preventDefault(){}});
        expect(setCategories).not.toHaveBeenCalled();
    })
    
    test('Debe de llamar el setCategories y limpiar la caja de texto', () => {
        const value = 'Hola';
        //busca el input
        const input =wrapper.find('input');
        //simula el cambio en el input, pasando metodo y el valor
        input.simulate('change',{target : {value}});
        // simula el envio del formulario, junto con el metodo
        wrapper.find('form').simulate('submit',{preventDefault(){}});
        //verifico que se ejecute 1 vez.
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1);
        expect(setCategories).toHaveBeenCalledWith( expect.any(Function));
        //obtengo una referencia a las propiedades del input
        const props = wrapper.find("input").props();
        // verifico que el valor del inputValue se haya reseteado
        expect(props.value).toBe('');
    });
})
