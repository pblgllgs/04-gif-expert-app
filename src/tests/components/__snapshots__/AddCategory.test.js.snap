// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> Debe mostrar <AddCategory /> correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    placeholder="Ingresa una categoria"
    type="text"
    value=""
  />
</form>
`;
